doc/rst/developer/bestPractices/ContributorInfo.rst

```diff
@@ -854,30 +888,69 @@ Pull request guidance
-    1000-line PR is acceptable, while a set of new tests introduced as a
-    1000-line PR is not.
+    the developer and user community.
+
+.. _Pull request description:
+
+Pull request description
+++++++++++++++++++++++++
+
+* Generally, try to come up with a single-line headline of 75 characters at
+  most to describe the change.  Think of this as a statement of what the PR
+  accomplishes, ideally with a user perspective (as opposed to the "what I did
+  to the code" perspective).
+
+* Depending on the magnitude of the change, write a short paragraph to longer
+  set of paragraphs describing the change (again, focusing on the user
+  experience, at least in the initial paragraphs... why would a Chapel user
+  care that you merged this?  Or if it's not user-facing, why would a developer
+  care).
+
+* If it's truly complex such that you think the code changes themselves warrant
+  describing in the PR (because they're complex or clever, say), additional
+  paragraphs talking about the code changes / approach taken can be nice (but
+  aren't strictly required).
+
+* Depending on the complexity of the PR, it can be reassuring to reviewers to
+  denote how you have tested the PR either in the PR description or an initial
+  comment.
+
+It's not uncommon for the contents of the PR description to evolve over the
+lifetime of the PR and its review.  Initially it should help the person doing
+the code review understand what's being done (and potentially why).  Then, when
+the code review is done, make sure the text is still accurate.
 
 .. _Final merge message:
 
 Final merge message
 +++++++++++++++++++
 
-- start with a single topic line with at most 75 characters
-- then have a blank line
-- then have a more detailed explanation including motivation for the
-  change and how it changes the previous behavior
-- use present tense (e.g. "Fix file iterator bug")
-- manually wrap long lines in the explanation to 75 or 80 characters
+When merging, copy and paste the PR description into the GitHub merge dialogue
+box, include the title of the PR which should already be there.
+
+It is good practice to add mention of the reviewer(s) at the end of the PR.
+This is typically formatted in brackets:
 
-It is good practice for the pull request title to match the topic line, and the
-detailed explanation to match the pull request description.
+.. code-block:: bash
+
+    [Reviewed by @reviewer1 and @reviewer2]
 
 .. _Git history is clear:
 
 Git history is clear
 ++++++++++++++++++++
 
+In general, having logical commits with meaningful commit messages is helpful
+for maintaining a clean git history. This is particularly important for
+critical or complex code changes. Depending on how critical or complex your
+changes are, it may be a good idea to do an interactive rebase to squash any
+non-meaningful commits:
+
+.. code-block:: bash
+
+    git fetch upstream
+    git rebase -i upstream/master
+
```

Comment:
This is impractical when I have hundreds of commits I brought from master with `git pull chapel-lang master`. Hopefully this will not be the case for the reader!

Instead/in addition I would like to suggest using the "squash and merge" option on GitHub's "merge" button when the developer does not value preserving individual commits and especially when there are merge commits.

Reply:
P.S. At this point this .rst is already filled with so many useful tips that I salute any reader who will get this far and actually use what we have written.